src/main/java/org/wso2/keycloak/client/KeycloakClient.java

```diff
@@ -0,0 +1,933 @@
+                KeycloakConstants.CLIENT_ENDPOINT;
+        String[] scope = ((String) oAuthApplicationInfo.getParameter(KeycloakConstants.TOKEN_SCOPE)).split(",");
+        Object tokenGrantType = oAuthApplicationInfo.getParameter(KeycloakConstants.TOKEN_GRANT_TYPE);
+        CloseableHttpClient httpClient = HttpClientBuilder.create().build();
+        //BufferedReader reader = null;
+        Map<String, Object> paramMap = new HashMap<String, Object>();
+        try {
+            // Create the JSON Payload that should be sent to OAuth Server.
+            String jsonPayload = createJsonPayloadFromOauthApplication(oAuthApplicationInfo, paramMap);
+            String accessToken = getAccessToken();
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Payload to create a new client : %s for the application %s", jsonPayload,
+                        clientName));
+            }
+            HttpPost httpPost = new HttpPost(registrationEndpoint);
+            httpPost.setEntity(new StringEntity(jsonPayload, KeycloakConstants.UTF_8));
+            httpPost.setHeader(KeycloakConstants.HTTP_HEADER_CONTENT_TYPE, KeycloakConstants.APPLICATION_JSON);
+            httpPost.setHeader(KeycloakConstants.AUTHORIZATION, KeycloakConstants.AUTHENTICATION_BEARER + accessToken);
+
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Invoking HTTP request to create new client in Keycloak for the application %s",
+                        clientName));
+            }
+            HttpResponse response = httpClient.execute(httpPost);
+            int statusCode = response.getStatusLine().getStatusCode();
+
+            // If successful a 201 will be returned with no body
+            if (HttpStatus.SC_CREATED == statusCode) {
+                String clientSecret = getClientSecret(clientName);
+                JSONObject clientInfoJsonObject = getClientById(clientName);
+                oAuthApplicationInfo = createOAuthAppInfoFromResponse(clientInfoJsonObject);
+                oAuthApplicationInfo.addParameter(KeycloakConstants.TOKEN_SCOPE, scope);
+                oAuthApplicationInfo.addParameter(KeycloakConstants.TOKEN_GRANT_TYPE, tokenGrantType);
+                oAuthApplicationInfo.addParameter(KeycloakConstants.CLIENT_SECRET, clientSecret);
+                oAuthApplicationInfo.setClientSecret(clientSecret);
+                return oAuthApplicationInfo;
+            } else {
+                handleException(String.format("Error occured while registering the new client in Keycloak. " +
+                        "Response : %s", statusCode));
+            }
+        } catch (UnsupportedEncodingException e) {
+            handleException(KeycloakConstants.ERROR_ENCODING_METHOD_NOT_SUPPORTED, e);
+        } catch (IOException e) {
+            handleException("Error while reading response body", e);
+        } finally {
+            closeResources(null, httpClient);
```

Comment:
we can use try-resource instead.

Reply:
I will try this.

Reply:
In fact, the closable resources defined in try are final. I cannot initialize my buffered reader there since I need to reassign the entity value from HTTP response to it. To keep the code consistent, I will keep using the finally block.

Reply:
I have used multiple try blocks and accommodated this recommendation.